sql/inspect: fix AS OF SYSTEM TIME span bounds in index checks

spilchen opened the pull request:
INSPECT's index consistency checks could miss spans when using AS OF SYSTEM TIME if those spans had no data at the current time but did at the historical timestamp. This was due to SpanToQueryBounds() using the current time for KV scans instead of the historical time.

This commit:
- Adds an asOf parameter to SpanToQueryBounds() for historical scans
- Updates INSPECT to pass the query's timestamp
- Keeps existing behavior for TTL and tests by passing an empty timestamp

The issue was uncovered while converting TestScrubIndexMissingIndexEntry to INSPECT. The test is now TestMissingIndexEntryWithHistoricalQuery and validates correctness of historical scans.

Informs: #155485
Epic: CRDB-55075

Release note (bug fix): INSPECT now correctly checks index consistency at the historical timestamp when using AS OF SYSTEM TIME, even for spans with no current data.

rafiss reviewed (approved):
nice find!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155837#-:-OcBJ2eg3k-_00jB5Zyc:b-jee1py)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 360 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155837#-OcBIHlJ2VKoFT5gVEtQ:-OcBIHlJ2VKoFT5gVEtR:b-mkls7c) ([raw file](https://github.com/cockroachdb/cockroach/blob/675cd1f80edadf6589252f47fbcebab3d766cb3d/pkg/sql/ttl/ttljob/ttljob_processor.go#L360)):*
> ```Go
> 				span,
> 				&alloc,
> 				hlc.Timestamp{}, // TTL doesn't use AS OF SYSTEM TIME
> ```

i'm wondering if this actually was a bug in the TTL job as well. the SELECT phase of TTL uses `ttlSpec.AOSTDuration` to pick the timestamp at which all SELECT queries should run. no need to fix it with this PR, but i did want to raise the topic so we can figure out what to do going forwards.

regardless of whether we need the fix or not, perhaps this comment can be clarified a bit further, since TTL does use AOST.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/inspect: fix AS OF SYSTEM TIME span bounds in index checks

INSPECT's index consistency checks could miss spans when using AS OF SYSTEM
TIME if those spans had no data at the current time but did at the historical
timestamp. This was due to SpanToQueryBounds() using the current time for KV
scans instead of the historical time.

This commit:
- Adds an asOf parameter to SpanToQueryBounds() for historical scans
- Updates INSPECT to pass the query's timestamp
- Keeps existing behavior for TTL and tests by passing an empty timestamp

The issue was uncovered while converting TestScrubIndexMissingIndexEntry to
INSPECT. The test is now TestMissingIndexEntryWithHistoricalQuery and validates
correctness of historical scans.

Informs: #155485
Epic: CRDB-55075

Release note (bug fix): INSPECT now correctly checks index consistency at the
historical timestamp when using AS OF SYSTEM TIME, even for spans with no
current data.

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155837)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 360 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155837#-OcBIHlJ2VKoFT5gVEtQ:-OcBlO2P-LUVmuMBfaAP:b-2rl8qd) ([raw file](https://github.com/cockroachdb/cockroach/blob/675cd1f80edadf6589252f47fbcebab3d766cb3d/pkg/sql/ttl/ttljob/ttljob_processor.go#L360)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i'm wondering if this actually was a bug in the TTL job as well. the SELECT phase of TTL uses `ttlSpec.AOSTDuration` to pick the timestamp at which all SELECT queries should run. no need to fix it with this PR, but i did want to raise the topic so we can figure out what to do going forwards.

regardless of whether we need the fix or not, perhaps this comment can be clarified a bit further, since TTL does use AOST.
</blockquote></details>

Yes, this is a good point to raise. I wasn't aware that TTL used an ASOF timestamp. I'll update the comment. 

I think it’s fine not to pass a timestamp for TTL; in theory we could miss rows that were already deleted, but TTL only needs spans that currently contain rows. It doesn’t have any work to do on data that no longer exists.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155837)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 360 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155837#-OcBIHlJ2VKoFT5gVEtQ:-OcBrIYz9A9feTYs47hs:bwv7nkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/675cd1f80edadf6589252f47fbcebab3d766cb3d/pkg/sql/ttl/ttljob/ttljob_processor.go#L360)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Yes, this is a good point to raise. I wasn't aware that TTL used an ASOF timestamp. I'll update the comment. 

I think it’s fine not to pass a timestamp for TTL; in theory we could miss rows that were already deleted, but TTL only needs spans that currently contain rows. It doesn’t have any work to do on data that no longer exists.
</blockquote></details>

ah, nice, that makes sense. 

i wonder if there's a potential improvement to be had by using AOST here for TTL. in theory, maybe that would avoid some contention with foreground traffic when computing the query bounds.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155837)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 360 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155837#-OcBIHlJ2VKoFT5gVEtQ:-OcCRXEiCLzOqYUV5liK:b-jymxqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/675cd1f80edadf6589252f47fbcebab3d766cb3d/pkg/sql/ttl/ttljob/ttljob_processor.go#L360)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

ah, nice, that makes sense. 

i wonder if there's a potential improvement to be had by using AOST here for TTL. in theory, maybe that would avoid some contention with foreground traffic when computing the query bounds.
</blockquote></details>

Yeah, good point. I opened this tracking issue to handle this performance improvement for TTL: https://github.com/cockroachdb/cockroach/issues/155871


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
